pylint/checkers/base.py

```diff
@@ -157,6 +161,7 @@ class CamelCaseStyle(NamingStyle):
 class PascalCaseStyle(NamingStyle):
     """Regex rules for PascalCase naming style."""
 
+    TYPE_VAR_RGX = re.compile(r"_*[^\W\da-z]*_?([^\W\da-z_][^\WA-Z_]+)*(_co(ntra)?)?$")
```

Comment:
I tested the changes against Home Assistant and noticed that even the `PascalCase` style doesn't seem to be perfect just yet. For example `CALLABLE_T` would be allowed currently.

As there aren't any formal guidelines yet, I looked at https://github.com/python/typeshed to see what they are using. This is the regex test list I came up with
```python
_TypeVar
_TBE
_ActionT
_MessageT
AnyStr
_CallableT
_T
_KT
_VT
DeviceTypeT

_T_co
_T_contra
_ActionT_co
_ActionT_contra

# These should NOT match!
DICT_T
CALLABLE_T
ENUM_T
DeviceType
_StrType
```

Some observations
* It's common for TypeVar names to be just one or two chars (at most 3). In these cases all are capitalized.
  `T`, `_KT`
* If the name is a word or multiple words, usually (but not always) it has a `T` suffix.
* `_co` and `_contra` mentioned in PEP 484 is used combined with the first two rules.
* In rare cases `Type` was added as suffix, instead of `T`. `4x - Type` vs `47x - T` in `typeshed`. I would lean towards preventing `Type` as suffix. Only something like `DeviceTypeT` should be allowed. That way we can somewhat unify the TypeVar naming.
* Uppercase names with `_T` suffix should not be allowed. Those should be replaced,
  i.e. `CALLABLE_T` -> `CallableT`

This is the regex I came up with
```py
    TYPE_VAR_RGX = re.compile(
        r"^_{0,2}(?:[^\W\da-z_]+|(?:[^\W\da-z_][^\WA-Z_]+)+T?(?<!Type))(?:_co(?:ntra)?)?$"
    )
```

Reply:
Thanks for the additional work you did here @cdce8p! Much appreciated as always.

This sounds like a very well-thought out regex pattern and list of matches. I checked on https://regex101.com and this works perfectly. I'll go ahead and work on a new version of this PR with this pattern.

Reply:
I used regex101.com myself 😄
We could debate if we truly want to disallow the `Type` suffix. The more I think about it the less certain I am. What to you do think?

Reply:
I came up with one more: what about `T_GoodName`? It is part of the test cases I wrote right now, but no longer passes with this regex. It also doesn't feature in your list.

Reply:
> I used regex101.com myself 😄
> We could debate if we truly want to disallow the `Type` suffix. The more I think about it the less certain I am. What to you do think?

It feels a bit arbitrary, but if people want to keep using it they can easily overwrite the regex pattern? I think it does make sense.

Reply:
> I came up with one more: what about `T_GoodName`? It is part of the test cases I wrote right now, but no longer passes with this regex. It also doesn't feature in your list.

Wasn't it that I suggested it originally? I don't remember anymore.
I've not seen that pattern in typeshed though, only with the Uppercase variant in Home Assistant. I lean towards saying it's invalid. `GoodNameT` would be more consistent.

Reply:
I wanted to educate myself about typvar with variance, but I did not find the time. I'm going to trust you on this one 😄 I think it's a rather large and non opinionated set of possible names and we can always restrict it later when/if a consensus is reached in the python community.

Reply:
> > I used regex101.com myself 😄
> > We could debate if we truly want to disallow the `Type` suffix. The more I think about it the less certain I am. What to you do think?
> 
> It feels a bit arbitrary, but if people want to keep using it they can easily overwrite the regex pattern? I think it does make sense.

Not sure how many would actually choose to overwrite it but you're right, it's at least an option. Nevertheless we have an opportunity to set a good standard here that shouldn't be discarded.

There are a few small arguments in favor of keeping it, at least for now.
* The `Type` suffix is much more common with Type Aliases, e.g. `Union[..., ...]`.
* The suffix `T` could be considered established in typeshed.
* And yes, we can always change it later.

-> Let's leave it then (for now).